clang/lib/3C/ConstraintVariables.cpp

```diff
@@ -1109,10 +1121,43 @@ FunctionVariableConstraint::FunctionVariableConstraint(const Type *Ty,
   }
 
   // ConstraintVariable for the return.
-  ReturnVar = FVComponentVariable(RT, RTIType, D, RETVAR, I, Ctx, &N, ReturnHasItype);
-  int GenericIdx = ReturnVar.ExternalConstraint->getGenericIndex();
-  if (GenericIdx >= 0)
-    TypeParams = std::max(TypeParams, GenericIdx + 1);
+  ReturnVar = FVComponentVariable(RT, RTIType, D, RETVAR, I, Ctx,
+                                  &N, true, ReturnHasItype);
+
+  // Locate the void* params for potential generic use
+  std::vector<int> Voids;
+  if(ReturnVar.ExternalConstraint->isVoidPtr() &&
+      !ReturnVar.ExternalConstraint->getIsGeneric()) {
+    Voids.push_back(-1);
+  }
+  for(int i=0; i < ParamVars.size();i++) {
+    if(ParamVars[i].ExternalConstraint->isVoidPtr() &&
+        !ParamVars[i].ExternalConstraint->getIsGeneric()) {
+      Voids.push_back(i);
+    }
+  }
+  // Strategy: If there's one void*, turn this into a generic function.
+  // Otherwise, we need to constraint the void*'s to wild
```

Comment:
There need to be constraints somewhere that observe code like this, and decide not to make the function generic:
```
int foo(void *p) {
  int x = (int)p;
  return x+1;
}
```
This unsafe cast should force the `void *` to be treated non generically. Is that happening? For that matter, if we have
```
int bar(void *q) {
  return foo(q);
}
```
then `q` should also be non-generic because `foo` solves, above, to be non-generic.

Reply:
This may call for more complex handling. I had hoped to rely on constraints that linked internal and external. In that case, it would be rewritten generic, but with `p` constrained to an array or marked wild. Similar for `q`. Perhaps on rewrite I can check that `p` is not subtyped? I might have to make sure the constraints link generic variables?

Reply:
I don't think `foo` should be rewritten at all: It's not generic, it's wild. I imagine the cast to `int` will add a wild constraint, and that constraint will propagate, so you have the information you need to not rewrite it, yes? 

I guess the key is to avoid adding a wild constraint when you have something like 
```
_Itype_for_any(T) void *malloc(...) : itype(_Array_type(T));
void *foo(int x) {
  void *p = malloc(x); // don't add cast to WILD on p
  return p;
}
```
but adding it in other cases, e.g., 
```
void *bar(int x) {
  void *q = (void *)1; // do add cast to WILD on q
  return q;
}
```
?

Reply:
For the first example above (`int x = (int)p;`), Utils.c:isCastSafe has a comment saying that casting to a non-pointer is safe. If I change it to an `int *` it behaves as you requested. The rest is something we're working on now.